### DISCUSSION TITLE
Why we use q,k,k_max = (64, 64, 32) / (64, 128,128) / (64, 128, 2e16) for sm80?

### DISCUSSION THREAD

**Discussion by ZhangDY-6483** [facebookresearch/xformers#847] [Ideas]
In generate_kernel.py, we generate one specific arch's kernel with some fixed shapes, such as q,k,k_max = (64, 64, 32) / (64, 128,128) / (64, 128, 2e16) for sm80. I'm not sure why these three shapes are selected but not others?
(Did we emulate all the possible shape and test the performance of them and select the best combination? )
![image](https://github.com/facebookresearch/xformers/assets/64682152/007ac4fb-3a83-42dd-8a55-65f81b563671)

**Comment by danthe3rd** [comment]
Hi,
You might get better performance if you know ahead of time the embedding dimension and hardcode it in the kernel indeed.
However, there is a tradeoff because pre-building a lot of kernels will take a lot of space and compilation time.

**Reply by ZhangDY-6483** [reply]
Hi, thanks for your reply ~ And why we select three fixed shape as patterns during generating kernel (these three shapes are slected by xformers):  xformers/csrc/attention/cuda/fmha/generate_kernels.py ?

**Reply by danthe3rd** [reply]
The sizes we select are aligned with what researchers usually use, and what is best handled by the kernel

**Reply by ZhangDY-6483** [reply]
Got it ! Thank u very much~
